[WFCORE-4394] Enhanced audit logging

justinmcook opened the pull request:
Jira: https://issues.jboss.org/browse/EAP7-699

justinmcook commented:
Hi @jamezp please review, in particular about the integration with [WFCORE-4355](https://issues.jboss.org/browse/WFCORE-4355)

jamezp commented:
Looks go do me @justinmcook. I just want to note too if you think the API for WFLY-4355 will make this more difficult it's not required.

honza-kasik commented:
Hi @justinmcook , thank you! I took a look at this, and I think this is _a lot_ of features for a single RFE. We have to ensure, that all aspects mentioned in this analysis will be implemented and what is even more important: that, they will be tested. From a brief look at this I could identify several areas this could be split into:

- RFC 5424/RFC 3164 accordance
- User formatted strings
- `@Audit` annotation
- Criteria filtering
- ... (each point basically describes enough features for one RFE)

I just want to express my concerns and in my opinion, finer granularity would be beneficial here.

@jamezp @justinmcook  Is it a good idea to present one RFE this big? Is this just an analysis for one big umbrella RFE and there will be more smaller ones? Am I missing something? :)

justinmcook commented:
@jamezp I think using the API won't make it more difficult, or, at worst, just slightly more difficult

justinmcook commented:
@honza-kasik It is somewhat large, and I think it was briefly discussed about splitting the RFE into smaller ones, if you think that would be beneficial then I can look into doing so and keeping this/ELY-1712 as the umbrella issue/analysis

jamezp commented:
@justinmcook Just an FYI I've submitted a [PR](https://github.com/wildfly/wildfly-core/pull/3700) for WFLY-4355. Hopefully it will end up looking like this and will be usable for you. If you feel there needs to be any changes please let me know.

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+*** include meta
+*** facility values
+*** priority mapping from facility and severity values, in accordance with https://tools.ietf.org/html/rfc5424#page-9[RFC 5424 Section 6.2.1]
+** The following attributes will also be available, but should be discarded if timeQuality is not specified:
```
How this will be done? Using _requires_ relationship in model?

justinmcook replied:
I was thinking to have these available as methods within the parent method, such that they would only ever be available if the parent is included. I.e. (...).timeQuality(...).setTimeZone(...).build().(...)

honza-kasik replied:
How it will be done from users point of view? I don't think, that user will manipulate with source code. Or am I wrong? Sorry if I misunderstood you.

justinmcook replied:
Sorry, I misunderstood originally. Yes, I would use the requires relationship

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+*** hostname (?)
+*** ssl-context
+** Additional attributes which will need to be added are:
+*** include timeQuality
```
I see `timeQuality` is name used in RFC 5424. What will be the name of this attribute in WF model?

justinmcook replied:
I plan to use the same names as within the RFC, so, for example, timeQuality will be called timeQuality within the model.

honza-kasik replied:
From what I know (and the check-up with Martin), the preferred names for model are not camel case variants but names which has words separated with dashes. Like this: `time-quality`.

justinmcook replied:
I was thinking about within the code. Within the model, I would use the dashes version of the word like in your example

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+*** port
+*** transport type
+*** message format
+*** hostname (?)
```
Why the question mark? I see, that we already have `host-name` attribute in model (`/subsystem=elytron/syslog-audit-log=*`)

justinmcook replied:
Mistake on my part, I added the question as a reminder to look into this more, relating to a support issue, but forgot to remove the question mark and so will do so in the next update

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,283 @@
+* Support for RFC 5424 should be enhanced, according to the following parameters, all of which should be added to the ```syslogAuditLogType``` in the Elytron schema
+// Could not find information about this in legacy
+** A ```syslog-format``` attribute should be added which supports a value of ```RFC5424```
+** Some of the parameters needed for RFC 5424 already exist in the Elytron codebase, which are the following:
````
Do they also exists in WildFly model?

justinmcook replied:
Yes, they are currently available in the Elytron subsystem AuditResourceDefinitions

honza-kasik replied:
Thanks, and are they available to user?

justinmcook replied:
The attributes set by the user are available, I will clarify in the proposal which ones are available to the user (i.e. TCP) and which ones are internal/based off of events (i.e. severity)

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,283 @@
+
+=== Hard Requirements
+
+* Support for RFC 5424 should be enhanced, according to the following parameters, all of which should be added to the ```syslogAuditLogType``` in the Elytron schema
````
This RFE targets Wildfly. Where in the WildFly will be those parameters added? By parameters you mean attributes?

honza-kasik replied:
How and where can users access them?

justinmcook replied:
These attributes will be available under the 'syslog-audit-log' on the CLI

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,283 @@
+*** sysUpTime
+*** language
+
+* Support for RFC 3164 should be added, based off of the implementation for RFC 5424 above, as such should also be added to the ```syslogAuditLogType``` in the Elytron schema
````
I see, that RFC 5424 obsoletes RFC 3164. Do we want to support both? Also, again, where in WildFly (from user POV) will this be added?

justinmcook replied:
I added this based on other audit logging in WildFly, such as [tamper-detecting-audit-log](https://github.com/kabir/tamper-detecting-audit-log/blob/1db363a0abb5599a88d7043eb728e8bf0791757c/src/main/java/org/jboss/audit/log/TcpSyslogHandler.java#L163), having support for both. I wasn't sure however on if this should be ultimately included, what are your thoughts on including it vs not including it? As above, it would be added to the Elytron subsystem in WildFly-Core

honza-kasik replied:
IMHO this needs further investigation. Do customers use syslog defined by one or another RFC in their infrastructure? Would they welcome both formats?

justinmcook replied:
I will discuss this with support

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+** The implementation of the RFC 3164 message format can be implemented based on the https://github.com/kabir/tamper-detecting-audit-log/blob/master/src/main/java/org/jboss/audit/log/TcpSyslogHandler.java[tamper-detecting-audit-log]
+
+* Support for user formatted strings should be added
+** Currently, this feature does not have a way to be implemented
```
Will be this somehow improved? Can you please elaborate?

justinmcook replied:
This is something that was requested in the RFC, but I'm unsure what it would refer to exactly. I included it in the proposal for feedback on what should be added for this capability

honza-kasik replied:
What I am looking for is description how can user benefit from this new "feature". IMHO it doesn't matter how it will be implemented, rather than how the user can benefit from it.

justinmcook replied:
I will discuss this with support and @darranl

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+* Support for user formatted strings should be added
+** Currently, this feature does not have a way to be implemented
+
+* There should be the ability to specify a Java class should be logged by using an @Audit Java annotation, similarly to what was available in Picketbox
```
Can you please elaborate on this? This analysis will be also used as a resource for documentation and it is not clear for me how the `@Audit` annotation should be used.

justinmcook replied:
This is one of the feature parity with legacy auditing parts. As far as I was able to tell, in PicketBox, the @Audit annotation can be added onto a class to signify that instances of that class being called should be logged. For the resource in Picketbox, see: https://developer.jboss.org/wiki/PicketBoxSecurityAnnotations#jive_content_id_SecurityAudit_Annotation

honza-kasik replied:
Thank you. I know, it seems clear to you and it might seemingly be clear to me, but this analysis must be comprehensible for a new comer. For someone who have never used features provided in legacy audit log configuration. For example, it is possible, that technical writers who will be processing this document will not know anything about it, because they have no experience.

justinmcook replied:
I'll add the details from this comment to the proposal for it is better clarified

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+
+* There should be the ability to specify a Java class should be logged by using an @Audit Java annotation, similarly to what was available in Picketbox
+
+* There should be the ability to configure routing, and create a custom filter allowing only a selection of events to be logged, based on several criteria, listed in the below table
```
This is about custom message format or introducing some filtering system into what will be logged?

justinmcook replied:
This refers to a filtering system

honza-kasik replied:
In that case, can you please explain in this analysis how the user will configure it and how it will be introduced to the WildFly model?

justinmcook replied:
Sure, I'll add it with the next update

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,283 @@
+** In each case, the user should be able to specify the criteria, the value(s), and a list of configuration changes from the default values, such as a different destination
+** Specifying the criteria, the value(s), and the configuration changes should be done as follows:
+*** When the programmatic approach is used, a method call should be available to the endpoint which will accept a given criteria, the corresponding values, and the filtering options
+*** When the markup approach is used, the inputs should be specified via attributes ```criteria="supported_criteria"```, ```values="respective_value_input"```, ```filtering="option1=change,option2=change,..."```.
````
Where will be the markup approach used? How can user use it?

justinmcook replied:
By markup, I refer to using the XML files, i.e. standalone.xml. This would be similiar to the management audit logging described in https://docs.jboss.org/author/display/WFLY8/Audit+logging and also similar to the following code snippet:
`<audit-logging><file-audit-log format="JSON" name="local-audit" path="audit.log" relative-to="jboss.server.log.dir"/></audit-logging>`

honza-kasik replied:
Thank you for explanation. However, editing of XML files no longer recommended way how to configure the server. Could you please instead of XML reference the model changes?

justinmcook replied:
Ah okay, I didn't realize this was no longer recommended (this is my first time modifying the model in detail). I'll change this section to reference the model changes

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+* There should be the ability to configure routing, and create a custom filter allowing only a selection of events to be logged, based on several criteria, listed in the below table
+** In each case, the user should be able to specify the criteria, the value(s), and a list of configuration changes from the default values, such as a different destination
+** Specifying the criteria, the value(s), and the configuration changes should be done as follows:
+*** When the programmatic approach is used, a method call should be available to the endpoint which will accept a given criteria, the corresponding values, and the filtering options
```
When can user use programmatic approach?

justinmcook replied:
This would be available to any Java program that is extending wildfly-elytron-audit, similarly to the Picketbox example in https://developer.jboss.org/wiki/PicketBoxAudit

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+
+|===
+
+* Additional audit events should be added for feature parity with Legacy Security Audit Logging. The following tables makes a comparison of Legacy Security Audit Logging features included in, not included in, and if the feature should be added to Elytron Audit Logging:
```
What do you mean by adding event? Where they will be added?

justinmcook replied:
By adding events I mean adding additional options for events that can be logged and they would coexist with the other event types in org.wildfly.security.auth.server.event and be added to the formatters

honza-kasik replied:
I am sorry, but I still don't understand what does that mean from users point of view.

justinmcook replied:
For the user, it would just affect the types of events that get logged when they look at their log. It would also tie into the filtering system above in having more options for the types of events that the user wishes to log/not log

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+| "hostname" attribute
+| No
+| https://access.redhat.com/support/cases/#/case/02188481[Case 02188481]
+| Currently supported in Elytron Audit Logging resulting in just needing the attribute to be added to the management configuration
```
I already see `host-name` attribute in syslog configuration in Elytron model. How this will be different?

justinmcook replied:
Originally I included this because of the support case, but, looking back at it, I think this is out of scope for this RFE and so I'll remove this section.

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+
+== Test Plan
+
+Tests for the new Audit Logging capabilities will be added.
```
Which tests and where they should be added?

justinmcook replied:
The tests will be added in the same package as where the code is added and will verify that the added Endpoints/formatters/etc work correctly, I will put additional detail for this in the next update

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+** An AysncAuditEndpoint should be added to WildFly-Core that uses this API to log events to a queue before they're writing to Stdout
+** JsonSecurityEventFormatter should be moved to WildFly-Core and modified to use the JsonFormatter in this API
+** The current FileAuditEndpoints, FileAuditEndpoint, PeriodicRotatingFileAuditEndpoint, and SizeRotatingFileAuditEndpoint
+should be moved to WildFly-Core
```
What changes will this introduce to model (if any)?

justinmcook replied:
I don't think this would introduce any changes to users other than additional Endpoints being available

honza-kasik replied:
Thanks! Where/How can user access them?

justinmcook replied:
These would be additional attributes in the model, I'll add in their names to the proposal

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,283 @@
+*** message format
+*** hostname (?)
+*** ssl-context
+** Additional attributes which will need to be added are:
```
Please include type of attributes which will be added (`String`, `int`...).

justinmcook replied:
Some of the attributes I won't know what class they would be until I start implementing it, but I'll define the ones I can

honza-kasik commented:
@justinmcook Hi Justin! I have made a small review of your analysis. WDYT?

justinmcook commented:
@honza-kasik Hi! Sorry I haven't responded to your comments yet, I plan to reply to them by the end of the day

honza-kasik commented:
@justinmcook Thank you again, Justin. I replied to some of your comments, yet I need to discuss few things which I do not know about so much with Martin (who has sickday today).

justinmcook commented:
Updated the PR to reflect the splitting of the task

honza-kasik commented:
@justinmcook Thank you for the update. I would prefer talking about changes introduced to the  elytron subsystem in the model -- newly introduced attributes and  operations which I can use to configure anything I need. Talking about model also allows us to talk about whether the attribute is required, its default value, capabilities and so on.

Also, I, as a  user use only supported ways of configuration such as CLI and/or web console, not editing XML files directly (this is an implementation detail which can change anytime). This analysis should describe changes made to the model. Changes which are available to use by user. Not implementation details and/or changes in private API.

fjuma commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,72 @@
+
+* The ability to configure for less reliability for increased speed should be added
+** This should be done by configuring the number of reconnect attempts to a syslog-server
+** This should be available as an attribute ```reconnect-attempts``` under ```syslog-audit-log``` in the Elytron subsystem
````
Will this new attribute have a default value?

justinmcook replied:
Updated the document to reference this

fjuma commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,72 @@
+
+=== Hard Requirements
+
+* Support for RFC 5424 should be enhanced, according to the following parameter,
```
Just a minor comment, it would be good to add links to the RFCs.

justinmcook replied:
Added the links

justinmcook commented:
@honza-kasik Updated the document

darranl commented:
As the analysis documents are analysis documents and not design documents I don't think it is always true that model changes can be fully defined, an engineer might have an idea how this will be represented in the model but this can evolve during design and implementation.  Regarding documentation by producing community documentation as the feature is merged we have then provided an account describing the end result of the work - at that stage the changes being made are definitely known.

I would also say the implementation detail if it is known is a valuable part of the analysis documents, as these pull requests are sent in these analysis documents are available to the wider engineering team who may also need an input on how a feature is to be developed - by omitting implementation details those engineers don't see anything to comment on the approach until the final pull requests are sent in which is too late in the process.

honza-kasik commented:
@justinmcook Thank you very much for the updated analysis! This look good from the QA point of view!

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,73 @@
+
+== Overview
+
+Currently, Audit Logging in Elytron does not have the same capabilities as Legacy Security. This task is to enhance Elytron Audit Logging to have the same capabilities as Legacy Security Audit Logging.
```
@justinmcook Just one last change. Can you please edit the overview so it reflects current state of this analysis?

justinmcook replied:
Edited the overview

honza-kasik commented:
Sorry @justinmcook , I missed the overview - can you please edit it? Otherwise, this looks good!

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,73 @@
+* The ability to configure for less reliability for increased speed should be added
+** This should be done by configuring the number of reconnect attempts to a syslog-server
+** This should be available as an attribute ```reconnect-attempts``` under ```syslog-audit-log``` in the Elytron subsystem
+** This would be an optional parameter and not have a default value
````
Just a very last thing. Wouldn't it be better if this has a default value of current count of reconnect attempts and making this a required attribute?

justinmcook replied:
Updated this section to still be optional but have a default value of 0

honza-kasik replied:
Thank you. Sorry, I meant "required" in sense, that it will have default value presenting current amount of reconnect attempts. Users won't have to set anything if they don't want to. If it will have default value of `0`, does it mean, that there will be no reconnect attempts made, or there will be infinite amount of them?

justinmcook replied:
By 0 I mean there would be no reconnect attempts

honza-kasik replied:
So how many reconnect attempts is done now?

justinmcook replied:
Currently it indefinitely attempts to reconnect

honza-kasik replied:
In that case, wouldn't be the value of `-1` best? If user wishes to change the behavior, he can set any positive number (thus swapping reliability for speed).

justinmcook replied:
Ya that would be better, I'll change it to -1

justinmcook replied:
Updated the description

honza-kasik commented:
@justinmcook I am sorry for last moment edits, but can you please edit the overview and WDYT about `recconect-attempts` comment?

justinmcook commented:
@honza-kasik Updating the overview now. For the reconnect-attempts: I think updating it to still be an optional attribute but with a default value would be best, because there would always have to be a specified number of reconnect attempts, but the attribute would only be needed if a user wished to customize that. On the same note, I will update syslog-format to also be an optional attribute, since it would default to RFC5424 and the attribute would only need to be added by a user if they wished to use RFC3164 (or any other RFC if any new ones are created in the future)

justinmcook commented:
@darranl @jamezp I was looking further into configuring the amount of reconnect attempts, and it looks like to be able to do this would require modifying the [TcpOutputStream class in jboss-logmanager](https://github.com/jboss-logging/jboss-logmanager/blob/master/ext/src/main/java/org/jboss/logmanager/ext/handlers/TcpOutputStream.java) for that the [RetryConnector](https://github.com/jboss-logging/jboss-logmanager/blob/master/ext/src/main/java/org/jboss/logmanager/ext/handlers/TcpOutputStream.java#L336) can take in the number of maximum attempts as an argument, rather than it currently being indefinite. Do you guys think making those modifications to jboss-logmanager would be worth the ability to configure the maximum number of reconnect attempts?

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
````diff
@@ -0,0 +1,76 @@
+* The ability to configure for less reliability for increased speed should be added
+** This should be done by configuring the number of reconnect attempts to a syslog-server
+** This should be available as an attribute ```reconnect-attempts``` under ```syslog-audit-log``` in the Elytron subsystem
+** The possible values for this parameter would be: -1 - signifying indefinite reconnect attempts,
````
@justinmcook From our discussion I understood that currently there is only on reconnect attempt made (and error is ignored). From my point of view we should keep backward compatibility and keep this as is by default. Can you please find out if current behavior is intended (if it's not a bug)?

cc @fjuma

justinmcook commented:
Updated the analysis document to reflect changes/discussions. In particular, the change of the default value for ```reconnect-attempts``` to 0 and noting that it only applies to UDP.

honza-kasik commented on elytron/ELY-1712-enhanced-audit-logging.adoc:
```diff
@@ -0,0 +1,80 @@
+0 - signifying no reconnect attempts, and any positive integer - signifying that nunmber of
+reconnect attempts
+** This would be an optional parameter and have a default value of 0
+** This parameter will only apply for logging to a syslog server with UDP, since TCP already has reconnections configured in the jboss-logmanager project
```
Thanks for the update @justinmcook! Is there a way how user can configure the reconnect attempts for TCP connection?

justinmcook commented:
Updated the analysis document to reflect discussions with @honza-kasik, in particularly changing the attribute ```reconnect-attempts``` to ```icmp-error-reconnect-attempts``` and updating its description

honza-kasik commented:
Thanks for the update @justinmcook. Please also add information why the attribute was named like this (ICMP error).

justinmcook commented:
Updated the analysis document explaining that reconnection will be attempted if there is an ```IOException``` and changed the attribute back to ```reconnect-attempts``` from ```icmp-error-reconnect-attempts``` to reflect there being no guarantee that the ```IOException``` being due to an ICMP error
